### PR TITLE
Properly handle unicode in safe_eval

### LINKED ISSUE #66943: Unicode breaks dict
##### SUMMARY

Ansible can't parse unicode correctly, when reading template

##### COMPONENT NAME

template

##### ANSIBLE VERSION

```
ansible 2.9.2
  config file = None
  configured module search path = [u'/home/home/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.5 (default, Aug  7 2019, 00:51:29) [GCC 4.8.5 20150623 (Red Hat 4.8.5-39)]
```

##### OS / ENVIRONMENT
 
CentOS Linux release 7.7.1908 (Core)

##### STEPS TO REPRODUCE

vars.yml

```yaml
some_var: "test"
```

templates/example.json.j2

```json
{
  "something": "test",
  "this-breaks-key": "This breaks the whole file, because of: ü§",
  "another-key": "{{ some_var }}"
}
```

```yaml
- hosts: test
  connection: local
  vars:
    reference: "{{ lookup('template', 'templates/example.json.j2') }}"
    delta: {}
  tasks:
    - name: Test | Get configuration changes
      set_fact:
        delta: "{{ delta | combine({item.key: item.value}, recursive=true) }}"
      loop: "{{ reference | dict2items }}"
```

##### EXPECTED RESULTS

The valid json file should be correctly parsed and unicode should not be a problem in 2020.

Everything works just as expected, when `unicode chars` are removed.

##### ACTUAL RESULTS

```json
{"msg": "dict2items requires a dictionary, got <type 'unicode'> instead."}
```

The underlying exception that causes this behavior:

```
Traceback (most recent call last):
  File "/Users/matt/projects/ansibledev/ansible/lib/ansible/template/safe_eval.py", line 142, in safe_eval
    compiled = compile(parsed_tree, expr, 'eval')
UnicodeEncodeError: 'ascii' codec can't encode characters in position 87-88: ordinal not in range(128)
```

### PR DESCRIPTION
**Pull request by mkrizek**
##### SUMMARY

Fixes #66943

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

`lib/ansible/template/safe_eval.py`

### CODE REVIEW

File: lib/ansible/template/safe_eval.py [hunk 1]
```diff
@@ -22,7 +22,9 @@
 import sys
 
 from ansible import constants as C
-from ansible.module_utils.six import string_types
+from ansible.module_utils._text import to_native
+from ansible.module_utils.common.text.converters import container_to_text
```
Comment (samdoran) [inline lib/ansible/template/safe_eval.py]:
```suggestion
from ansible.module_utils.common.text.converters import container_to_text, to_native
```

File: lib/ansible/template/safe_eval.py [hunk 2]
```diff
@@ -139,11 +141,15 @@ def generic_visit(self, node, inside_call=False):
     try:
         parsed_tree = ast.parse(expr, mode='eval')
         cnv.visit(parsed_tree)
-        compiled = compile(parsed_tree, expr, 'eval')
+        compiled = compile(parsed_tree, to_native(expr), 'eval')
```
Comment (samdoran) [inline lib/ansible/template/safe_eval.py:143]: Do we want errors with this?
```suggestion
        compiled = compile(parsed_tree, to_native(expr, errors='surrogate_or_strict'), 'eval')
```

**Review by samdoran — CHANGES_REQUESTED** [review]

File: lib/ansible/template/safe_eval.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/template/safe_eval.py:143]: I'm not sure if that makes sense. I want to look at this some more tomorrow. Today was hectic. 🤯

Comment (mkrizek) [reply lib/ansible/template/safe_eval.py:143]: I don't think we need `errors='surrogate_or_strict'` for that. It's `filename` argument for the `compile` function [0]:

> The filename argument should give the file from which the code was read; pass some recognizable value if it wasn’t read from a file ('<string>' is commonly used).

So the value is information only and we are fine with omitting "bad bytes" from the value.

[0] https://docs.python.org/3/library/functions.html#compile

File: lib/ansible/template/safe_eval.py (hunk 1 above)
Comment (mkrizek) [reply lib/ansible/template/safe_eval.py]: :+1:

File: lib/ansible/template/safe_eval.py (hunk 2 above)
Comment (samdoran) [reply lib/ansible/template/safe_eval.py:143]: Ok, that makes sense. Thanks for the great explanation.

**Comment by samdoran** [comment]
Working on a fix for that test failure. It's unrelated to this PR.

**Review by samdoran — APPROVED** [review]

**Pull request merged**